### PR TITLE
[Sle15] AY setup written in first stage

### PR DESCRIPTION
**Pull request by mchf**
* DNS supported
* /etc/hosts supported
* Routing supported

See https://github.com/yast/aytests-tests/pull/77 (DNS test)

### CODE REVIEW

File: src/lib/network/network_autoyast.rb [hunk 1]
```diff
@@ -106,24 +108,63 @@ def set_network_service
...
+    # @return [Boolean] true when configuration was present and loaded from the profile
+    def configure_hosts(write: false)
+      log.info("NetworkAutoYast: Hosts configuration")
+
+      # expected format for Hosts.Import is { "ip" => [list, of, names] }
+      hosts_config = (ay_host_section["hosts"] || {}).map do |host|
+        # we neet to guarantee order of the items here
+        [host["host_address"] || "", host["names"] || []]
+      end
+      hosts_config = hosts_config.to_h.delete_if { |k, v| k.empty? || v.empty? }
```
Comment (mvidner) [inline src/lib/network/network_autoyast.rb]: This non-trivial FromAY-like section should go into its own method (class?) and be covered with tests.

File: src/modules/DNS.rb [hunk 2]
```diff
@@ -691,6 +615,90 @@ def read_hostname_from_install_inf
...
+    # present => nil
+    def dhclient_set_hostname
+      SYSCFG_TO_BOOL[SCR.Read(path(".sysconfig.network.dhcp.DHCLIENT_SET_HOSTNAME"))]
+    end
+
+    # Reads value of WRITE_HOSTNAME_TO_HOSTS and translates it to boolean
+    #
+    # return {true, false, nil} "yes" => true, "no" => false, otherwise or not
+    # present => nil
+    def write_hostname_to_hosts
```
Comment (mvidner) [inline src/modules/DNS.rb]: A method named `write_*` that in fact calls `SCR.Read`?! I know it copies the sysconfig name but here I would prefix it with `should_` or something.

File: src/modules/DNS.rb [hunk 3]
```diff
@@ -381,65 +330,40 @@ def Write
...
-        HOSTNAME_PATH,
-        Ops.add(fqhostname, "\n")
-      )
-
-      create_hostname_link
-
-      Builtins.sleep(sl)
+      ProgressNextStage(_("Writing hostname...")) if gui
+      update_hostname
+      Builtins.sleep(0) if gui
```
Comment (mvidner) [inline src/modules/DNS.rb]: Ugh, what?!

File: src/modules/DNS.rb [hunk 4]
```diff
@@ -691,6 +615,90 @@ def read_hostname_from_install_inf
...
+        # @dhcp_hostname and @wrote_hostname can currently be nil only when
+        # not present in original file. So, do not add it in such case.
+        SCR.Write(
+          path(".sysconfig.network.dhcp.DHCLIENT_SET_HOSTNAME"),
+          @dhcp_hostname ? "yes" : "no"
+        ) if !@dhcp_hostname.nil?
+        SCR.Write(
+          path(".sysconfig.network.dhcp.WRITE_HOSTNAME_TO_HOSTS"),
+          @write_hostname ? "yes" : "no"
+        ) if !@write_hostname
```
Comment (mvidner) [inline src/modules/DNS.rb]: Bug.

File: src/lib/network/network_autoyast.rb (hunk 1 above)
Comment (mchf) [inline src/lib/network/network_autoyast.rb]: well this is almost everything what ```configure_hosts``` does

File: src/modules/DNS.rb (hunk 3 above)
Comment (mchf) [inline src/modules/DNS.rb]: hm, consequence of some massive updating and code shifting and so on. I'll drop it.

File: src/clients/save_network.rb [hunk 5]
```diff
@@ -259,16 +259,41 @@ def configure_dns
...
+      NetworkAutoYast.instance.configure_routing
+    end
+
+    def configure_nics
+      return if !Mode.autoinst
+
+      NetworkAutoYast.instance.configure_nics
+    end
+
     # It does an automatic configuration of installed system
```
Comment (mchf) [inline src/clients/save_network.rb]: upgrade scenario ignores network setup ... but i'm not sure about autoupgrade - will check

File: package/yast2-network.changes [hunk 6]
```diff
@@ -1,4 +1,12 @@
 -------------------------------------------------------------------
+Fri Jun  9 09:40:11 UTC 2017 - mfilka@suse.com
+
+- Moving network setup in AY into 1s stage
```
Comment (mchf) [inline package/yast2-network.changes]: currently not aware of any - the PBI contains no fate number 

@kobliha: is there any fate request?

File: src/lib/network/network_autoyast.rb (hunk 1 above)
Comment (mvidner) [inline src/lib/network/network_autoyast.rb]: Ok, now that the other configure_* methods are gone it makes less sense to split it off. Still, with all the "or empty" logic in here, tests are needed.

Also, how does this code relate to [Import in host_auto](https://github.com/mchf/yast-network/blob/f7eb35908554123042b8c570fa8f5f6517bea923/src/clients/host_auto.rb#L85)? Do we need to "guarantee item order" also there?

**Comment by mvidner** [comment]
does host_auto#Import need similar code like configure_hosts?

File: src/lib/network/network_autoyast.rb (hunk 1 above)
Comment (mchf) [inline src/lib/network/network_autoyast.rb]: Item order is most probably bad description here. I refer to "host_address" and "names". Each of these hash keys refer to a xml node. And this nodes can be ordered differently in the profile. We must guarantee that "host_address" is at first place.

**Comment by mchf** [comment]
yes, see https://github.com/yast/yast-network/blob/master/src/clients/host_auto.rb#L86

File: src/lib/network/network_autoyast.rb (hunk 1 above)
Comment (mvidner) [inline src/lib/network/network_autoyast.rb]: I think that this duplicates `Host.Import` . Is that one obsolete then?

Comment (mvidner) [inline src/lib/network/network_autoyast.rb]: Ah, sorry, `configure_submodule` does call `Import`

File: src/lib/network/network_autoyast.rb [hunk 7]
```diff
@@ -104,26 +106,40 @@ def set_network_service
...
+    # If the installer is running in 1st stage mode only, then the configuration
+    # is also written
+    #
+    # @param [Boolean] write forces instant writing of the configuration
+    # @return [Boolean] true when configuration was present and loaded from the profile
+    def configure_hosts(write: false)
+      log.info("NetworkAutoYast: Hosts configuration")
 
-      true
+      # expected format for Hosts.Import is { "ip" => [list, of, names] }
```
Comment (mvidner) [inline src/lib/network/network_autoyast.rb]: OK, and this must be documented *there*, in the API docs for Host.Import (not Hosts).

**Comment by mchf** [comment]
Writing in first stage is experimental feature from my POV and is used only when second stage is disabled. Othewise _auto clients are used in second stage. See AY integration tests for some more details.

File: src/lib/network/network_autoyast.rb [hunk 8]
```diff
@@ -206,16 +222,53 @@ def merge_routing(instsys_routing, ay_routing)
...
+
+    # Returns host section of the current AY profile
+    #
+    # Note that autoyast transforms the host's subsection
+    # into:
+    # {
+    #   hosts => [
+    #     # first <host_entry>
+    #     {
+    #       "hosts_address" => <ip>,
```
Comment (mchf) [inline src/lib/network/network_autoyast.rb]: ok

File: src/lib/network/network_autoyast.rb [hunk 9]
```diff
@@ -206,16 +222,53 @@ def merge_routing(instsys_routing, ay_routing)
...
 
       ay_profile = Profile.current
 
       return {} if ay_profile.nil? || ay_profile.empty?
-      return {} if ay_profile["networking"].nil?
+      ay_profile
+    end
+
+    # Returns networking section of current AY profile
+    def ay_networking_section
```
Comment (mchf) [inline src/lib/network/network_autoyast.rb]: regarding ```Host.Import``` see bellow

File: src/lib/network/network_autoyast.rb (hunk 7 above)
Comment (mchf) [inline src/lib/network/network_autoyast.rb]: ok

File: src/lib/network/network_autoyast.rb (hunk 9 above)
Comment (mchf) [inline src/lib/network/network_autoyast.rb]: in my POV "networking" section is used quite often so it deserved own method as a shortcut.

**Comment by mchf** [comment]
Thanks for the review

**Pull request merged**